src/invariant/InvariantTests.cpp

```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+                switch (mType)
+                {
+                case ACCOUNT:
+                    entry.data.account() = LedgerTestUtils::generateValidAccountEntry(5);
+                    break;
+                case TRUSTLINE:
+                    entry.data.trustLine() = LedgerTestUtils::generateValidTrustLineEntry(5);
+                    break;
+                case OFFER:
+                    entry.data.offer() = LedgerTestUtils::generateValidOfferEntry(5);
+                    break;
+                case DATA:
+                    entry.data.data() = LedgerTestUtils::generateValidDataEntry(5);
+                    break;
+                default:
+                    REQUIRE(false);
+                }
+
+                LedgerHeader lh;
+                LedgerDelta ld(lh, db, false);
+                auto ef = EntryFrame::FromXDR(entry);
+                ef->storeAdd(ld, db);
+                mAdded = true;
+            }
+        }
+        return ApplyBucketsWork::onSuccess();
+    }
+};
+
+class ApplyBucketsWorkDeleteEntry : public ApplyBucketsWork
+{
+  private:
+    uint32_t mFromLedgerSeq;
+    LedgerEntryType mType;
+    bool mDeleted;
+
+  public:
+    ApplyBucketsWorkDeleteEntry(
+            Application& app, WorkParent& parent,
+            std::map<std::string, std::shared_ptr<Bucket>> const& buckets,
+            HistoryArchiveState const& applyState,
+            uint32_t deleteFromLedgerSeq,
+            LedgerEntryType deleteType)
+        : ApplyBucketsWork(app, parent, buckets, applyState)
+        , mFromLedgerSeq{deleteFromLedgerSeq}
+        , mType{deleteType}
+        , mDeleted{false}
+    {
+    }
+
+    Work::State onSuccess() override
+    {
+        if (!mDeleted)
+        {
+            long long affectedRows = 0;
+            {
+                auto table = getTableName(mType);
+                auto& db = mApp.getDatabase();
+                auto prep = db.getPreparedStatement(
+                    "DELETE FROM " + table + " WHERE lastmodified= :v1 LIMIT 1");
+                auto& st = prep.statement();
+                st.exchange(soci::use(mFromLedgerSeq));
+                st.define_and_bind();
+                st.execute(true);
+                affectedRows = st.get_affected_rows();
+            }
+            assert(affectedRows <= 1);
+            if (affectedRows == -1)
+            {
+                throw std::runtime_error("Database operation unsupported");
+            }
+            mDeleted = (affectedRows == 1);
+        }
+        return ApplyBucketsWork::onSuccess();
+    }
+};
+
+class ApplyBucketsWorkModifyEntry : public ApplyBucketsWork
+{
+  private:
+    uint32_t mFromLedgerSeq;
+    LedgerEntryType mType;
+    bool mModified;
+
+    void maybeModifyAccountEntry()
+    {
+        auto& db = mApp.getDatabase();
+        std::string account;
+        {
+            auto prep = db.getPreparedStatement(
+                "SELECT accountid FROM accounts"
```

Comment:
you don't actually need this, standard load + check that last modified is what you expect does the same without introducing new SQL

Reply:
what I am advocating for is to avoid having SQL or soci code outside of the few classes that deal with it - it's fine to have helper methods exposed for tests only